drake/solvers/MobyLCP.cpp

```diff
@@ -0,0 +1,1268 @@
+void selectSubVec(const Eigen::MatrixBase<Derived>& in,
+                  const std::vector<unsigned>& rows,
+                  Eigen::VectorXd* out) {
+  const int num_rows = rows.size();
+  out->resize(num_rows);
+  for (int i = 0; i < num_rows; i++) {
+    (*out)(i) = in(rows[i]);
+  }
+}
+
+template <typename DerivedA, typename DerivedB, typename F>
+void transformVecElements(const Eigen::MatrixBase<DerivedA>& in,
+                          Eigen::MatrixBase<DerivedB>* out,
+                          F func) {
+  assert(in.cols() == 1);
+  assert(out->cols() == 1);
+  assert(in.rows() == out->rows());
+  for (int i = 0; i < in.rows(); i++) {
+    (*out)(i) = func(in(i), (*out)(i));
+  }
+}
+
+template <typename Derived>
+Eigen::SparseVector<double> makeSparseVector(
+    const Eigen::MatrixBase<Derived>& in) {
+  assert(in.cols() == 1);
+  Eigen::SparseVector<double> out(in.rows());
+  for (int i = 0; i < in.rows(); i++) {
+    if (in(i) != 0.0) {
+      out.coeffRef(i) = in(i);
+    }
+  }
+  return out;
+}
+
+template <typename Derived>
+Eigen::Index minCoeffIdx(const Eigen::MatrixBase<Derived>& in) {
+  Eigen::Index idx;
+  in.minCoeff(&idx);
+  return idx;
+}
+
+const double NEAR_ZERO = std::sqrt(std::numeric_limits<double>::epsilon());
+}
+
+namespace Drake {
+
+// Sole constructor
+MobyLCPSolver::MobyLCPSolver() : log_enabled_(false) {
+}
+
+void MobyLCPSolver::setLoggingEnabled(bool enabled) {
+  log_enabled_ = enabled;
+}
+
+std::ostream& MobyLCPSolver::LOG() const {
+  if (log_enabled_) {
+    return std::cerr;
+  }
+  return null_stream_;
+}
+
+void MobyLCPSolver::clearIndexVectors() const {
+  // clear all vectors
+  _all.clear();
+  _tlist.clear();
+  _bas.clear();
+  _nonbas.clear();
+  _j.clear();
+}
+
+bool MobyLCPSolver::solve(OptimizationProblem& prog) const {
+  // TODO ggould in principle it should be possible to render each of these as
+  // a LCP and merge the resulting problems, but I don't yet know how to do it.
+  assert(prog.getGenericConstraints().empty());
+  assert(prog.getGenericObjectives().empty());
+  assert(prog.getAllLinearConstraints().empty());
+  assert(prog.getBoundingBoxConstraints().empty());
+  assert(prog.getLinearComplementarityConstraint());
+
+  Eigen::VectorXd solution(prog.getNumVars());
+  bool solved = lcp_lemke_regularized(
+      prog.getLinearComplementarityConstraint()->getConstraint()->getM(),
+      prog.getLinearComplementarityConstraint()->getConstraint()->getq(),
+      &solution);
+  if (solved) {
+    prog.setDecisionVariableValues(solution);
+  }
+  return solved;
+}
+
+/// Fast pivoting algorithm for denerate, monotone LCPs with few nonzero,
+/// nonbasic variables
+bool MobyLCPSolver::lcp_fast(
+    const Eigen::MatrixXd& M, const Eigen::VectorXd& q, Eigen::VectorXd* z,
+    double zero_tol) const {
+  const unsigned N = q.rows();
+  const unsigned UINF = std::numeric_limits<unsigned>::max();
+
+  LOG() << "MobyLCPSolver::lcp_fast() entered" << std::endl;
```

Comment:
I think most of the more expensive logging lines are wrapped by if (log_enabled_), though I agree we're still doing string manipulation pointlessly in places where logging isn't enabled.  Until we figure out #1895, would just wrapping all of the LOG() invocations in log_enabled_ be sufficient?